scraper.py

```diff
@@ -0,0 +1,246 @@
+        await asyncio.sleep(args.interval)
+
+
+async def resolve_what(what: List[str], client: Client):
+    """
+    Resolve the list of "what" arguments on the command line to (tenant,timeline) tuples.
+    """
+    tenant_and_timline_ids: Set[Tuple[str, str]] = set()
+    # fill  tenant_and_timline_ids based on spec
+    for spec in what:
+        comps = spec.split(":")
+        if comps == ["ALL"]:
+            tenant_ids = await client.get_tenant_ids()
+            get_timeline_id_coros = [client.get_timeline_ids(tenant_id) for tenant_id in tenant_ids]
+            gathered = await asyncio.gather(*get_timeline_id_coros, return_exceptions=True)
+            assert len(tenant_ids) == len(gathered)
+            for tid, tlids in zip(tenant_ids, gathered):
+                for tlid in tlids:
+                    tenant_and_timline_ids.add((tid, tlid))
+        elif len(comps) == 1:
+            tid = comps[0]
+            tlids = await client.get_timeline_ids(tid)
+            for tlid in tlids:
+                tenant_and_timline_ids.add((tid, tlid))
+        elif len(comps) == 2:
+            tenant_and_timline_ids.add((comps[0], comps[1]))
+        else:
+            raise ValueError(f"invalid what-spec: {spec}")
+
+    return tenant_and_timline_ids
+
+
+async def main_impl(args, db: asyncpg.Pool, client: Client):
+    """
+    Controller loop that manages the per-timeline scrape tasks.
+    """
+
+    psid = await client.get_pageserver_id()
+    scrapedb_ps_id = f"{args.environment}-{psid}"
+
+    logging.info(f"storing results for scrapedb_ps_id={scrapedb_ps_id}")
+
+    active_tasks_lock = asyncio.Lock()
+    active_tasks: Dict[Tuple[str, str], asyncio.Event] = {}
+    while True:
+        try:
+            desired_tasks = await resolve_what(args.what, client)
+        except Exception:
+            logging.exception("failed to resolve --what, sleeping then retrying")
+            await asyncio.sleep(10)
+            continue
+
+        async with active_tasks_lock:
+            active_task_keys = set(active_tasks.keys())
+
+            # launch new tasks
+            new_tasks = desired_tasks - active_task_keys
+            for (tenant_id, timeline_id) in new_tasks:
+                logging.info(f"launching scrape task for timeline {tenant_id}/{timeline_id}")
+                stop_var = asyncio.Event()
+
+                async def task_wrapper():
+                    try:
+                        await timeline_task(
+                            args, scrapedb_ps_id, tenant_id, timeline_id, client, db, stop_var
+                        )
+                    finally:
+                        async with active_tasks_lock:
+                            del active_tasks[(tenant_id, timeline_id)]
+
+                assert active_tasks.get((tenant_id, timeline_id)) is None
+                active_tasks[(tenant_id, timeline_id)] = stop_var
+                asyncio.create_task(task_wrapper())
```

Comment:
Seems this `create_task` return value would have a `cancel(self)` method, did you already attempt that before going with the `stop_var` events? Would alleviate the need for `task_wrapper` and "select" in `timeline_task`...

Reply:
`.cancel()`causes a panic to be thrown at any `await` point inside `task_wrapper`.
Didn't like that we'd be `logging.exception()` in that case, makes the log noisy.

But I guess from a correctness perspective the script would handle it just fine.

Reply:
I understood the exception would only be thrown if there's an `await task` but may had been wrong. Unsure though what'll happen for unawaited cancelled tasks. Could of course just handle those silently.

Good to resolve either way.

Reply:
Hm, docs:

> Tasks can easily and safely be cancelled. When a task is cancelled, asyncio.CancelledError will be raised in the task at the next opportunity.

https://docs.python.org/3/library/asyncio-task.html#id4

Under that assumption, another argument for using the stop var is that our scraper resets the stats on every scrape (see the comment in `scrape_timeline`). So, if we just `task.cancel()`, we might lose the last scrape.